aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1387 @@
+    ///
+    /// This counter is monotonically increasing and updated in a successful try_abort.
+    /// It allows filtering fanned-out abort attempts when multiple workers executing
+    /// different transactions invalidate different reads of the same transaction.
+    /// Only one of these workers will successfully abort the transaction and perform
+    /// the required processing.
+    next_incarnation_to_try_abort: CachePadded<AtomicU32>,
+
+    /// Part of inner status state summarized as a single flag that can be read lock-free.
+    /// The allowed values are defined in the `dependency_flags` module.
+    dependency_shortcut: CachePadded<AtomicU8>,
+
+    /// Tracks the number of active stalls on this transaction.
+    ///
+    /// A transaction is considered "stalled" when this count is greater than 0.
+    /// Each add_stall increments this counter, and each remove_stall decrements it.
+    /// The status is "unstalled" when the counter returns to 0.
+    num_stalls: CachePadded<AtomicU32>,
+
+    /// Interface to manage the transaction execution queue.
+    ///
+    /// Allows adding or removing transactions from the execution queue based on
+    /// their status changes. Used when stalls are added/removed or when
+    /// a new incarnation is created.
+    execution_queue_manager: Arc<ExecutionQueueManager>,
+
+    /// Index of this transaction in the block.
+    txn_idx: TxnIndex,
+}
+
+impl ExecutionStatus {
+    pub(crate) fn new(
+        execution_queue_manager: Arc<ExecutionQueueManager>,
+        txn_idx: TxnIndex,
+    ) -> Self {
+        Self {
+            inner_status: CachePadded::new(Mutex::new(InnerStatus::new())),
+            next_incarnation_to_try_abort: CachePadded::new(AtomicU32::new(0)),
+            dependency_shortcut: CachePadded::new(AtomicU8::new(dependency_flags::DEFER)),
+            num_stalls: CachePadded::new(AtomicU32::new(0)),
+            execution_queue_manager,
+            txn_idx,
+        }
+    }
+
+    /// Attempts to transition a transaction from PendingScheduling to Executing state.
+    ///
+    /// This method is called by the scheduler when it selects a transaction for execution.
+    /// It only succeeds if the transaction is currently in PendingScheduling state.
+    ///
+    /// # Returns
+    /// - `Ok(Some(incarnation))` if the transition was successful, returning the incarnation number
+    /// - `Ok(None)` if the transaction is not in PendingScheduling state
+    /// - `Err(PanicError)` if there was an error during flag transition
+    pub(crate) fn try_start_executing(&self) -> Result<Option<Incarnation>, PanicError> {
+        let ret = self.inner_status.lock().try_start_executing();
+
+        if ret.is_some() {
+            // When status is PendingScheduling the dependency shortcut flag should be
+            // DEFER (default or set by abort under the inner status lock).
+            // It could also be SAFE if the transaction was previously executed
```

Comment:
Then why not have SAFE in expected values?

Reply:
this is some outdated incorrect comment - thanks for catching!
moreover, I think the lock should be held while updating the shortcut to make sure those previous value invariants hold up (we can optimize later if we wanted to) - fixed that as well.